proxy.go

```diff
@@ -275,16 +275,35 @@ func (rp *reverseProxy) serveFromCache(s *scope, srw *statResponseWriter, req *h
 	}
 
 	// Await for potential result from concurrent query
-	if userCache.AwaitForConcurrentTransaction(key) {
-		cachedData, err := userCache.Get(key)
-		if err == nil {
-			_ = RespondWithData(srw, cachedData.Data, cachedData.ContentMetadata, cachedData.Ttl, http.StatusOK)
-			cacheHitFromConcurrentQueries.With(labels).Inc()
-			log.Debugf("%s: cache hit after awaiting concurrent query", s)
-			return
+	transaction, err := userCache.AwaitForConcurrentTransaction(key)
+	if err != nil {
+		// log and continue processing
+		log.Errorf("failed to await for concurrent transaction due to: %v", err)
+	} else {
+		if transaction.State.IsCompleted() {
+			cachedData, err := userCache.Get(key)
+			if err == nil {
+				_ = RespondWithData(srw, cachedData.Data, cachedData.ContentMetadata, cachedData.Ttl, http.StatusOK)
+				cacheHitFromConcurrentQueries.With(labels).Inc()
+				log.Debugf("%s: cache hit after awaiting concurrent query", s)
+				return
+			} else {
+				cacheMissFromConcurrentQueries.With(labels).Inc()
+				log.Debugf("%s: cache miss after awaiting concurrent query", s)
+			}
 		} else {
-			cacheMissFromConcurrentQueries.With(labels).Inc()
-			log.Debugf("%s: cache miss after awaiting concurrent query", s)
+			if transaction.State.IsFailed() {
+				err = fmt.Errorf("%s: concurrent query failed", s)
+				respondWith(srw, err, http.StatusInternalServerError)
+				return
+			}
+			// We will assess if elapsed time is bigger than half of the max request timeout.
+			// This precondition halts execution
+			if transaction.ElapsedTime > s.user.maxExecutionTime/2 {
```

Comment:
after 2nd thought, this code seems risky because on cluster with long queries we could have that:
- the concurrent queries was not working (because of the equality on pointers you spot)
- people update their CHProxy version and starts to have "weird" errors
IMHO we should at least have a new conf to activate this behavior so that people know what they're doing when they activate it.
Another thing is that maybe maxExecutionTime/2 is to short because:
-  if people use the default CH 30 sec.
- Let's say you have an APP (or function module) with slow queries (like 25 sec)
- the end user refreshes it's UI and fire a new query after 5 sec
- and it will get an error even if the first query was able to finish.
Maybe the best would be something like s.user.maxExecutionTime - 2 sec if (maxExecutionTime>= 20 sec) or maxExecutionTime -1 sec if ....

Reply:
**For the first point** - I don't understand how the pointer or long query issue can be impacted by this predicate. Keep in mind that it's a branch in which result of concurrent query was either not found or exceeded **grace_time**. Now, it depends on the user how **grace_time** is defined.

We can indeed be conservative in choice and set the predicate to:
`transaction.ElapsedTime > s.user.maxExecutionTime -1`.

Reply:
Actually I have another proposition which would remove this condition. What if we removed `grace_time` configuration and use the value in code equal to `max_exec_time`?

Reply:
Regarding your first comment, my issue is that the grace_time is at 5 sec by default so someone not aware about this new feature could upgrade CHPRoxy and get weird KO if his workload is designed for slow (+ 20 sec) queries  (it's maybe an edge since I have not understanding of how people are using CH).

Regarding your second comment I agree with you that grace_time should be equal to max_exec_time and it would avoid this extra logic. Since the max_exec_time is optionnal, we should use a default value at 30 sec (same as CH), wdyt?